core/src/main/java/com/linecorp/armeria/client/endpoint/StickyStrategy.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2017 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.client.endpoint;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+import java.util.function.ToLongFunction;
+
+import com.google.common.hash.Hashing;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.Endpoint;
+
+/**
+ * An {@code EndpointSelector} strategy which implements sticky load-balancing using
+ * user passed {@code ToLongFunction<CLientRequestContext>} as hash calculator for
+ * consistent hashing.
+ */
+public class StickyStrategy implements EndpointSelectionStrategy {
+
+    private final ToLongFunction<ClientRequestContext> requestHasher;
+
+    private static final ToLongFunction<Endpoint> DEFAULT_ENDPOINT_HASHER = Endpoint::hashCode;
+
+    /**
+     * Creates a new {@link StickyStrategy}.
+     * with default hash function to hash a {@link ClientRequestContext} to a long
+     *
+     * @param requestHasher A default {@link ToLongFunction} of {@link ClientRequestContext}
+     */
+    public StickyStrategy(ToLongFunction<ClientRequestContext> requestHasher) {
+        this.requestHasher = requireNonNull(requestHasher);
+    }
+
+    /**
+     * Creates a new {@link StickySelector}.
+     *
+     * @param endpointGroup An {@link EndpointGroup}
+     * @return A new {@link StickySelector}
+     */
+    @Override
+    @SuppressWarnings("unchecked")
+    public EndpointSelector newSelector(EndpointGroup endpointGroup) {
+        return new StickySelector(requestHasher, endpointGroup, this);
+    }
+
+    private static final class StickySelector implements EndpointSelector {
+
+        private final ToLongFunction<ClientRequestContext> requestContextHasher;
+        private final EndpointGroup endpointGroup;
+        private final EndpointSelectionStrategy endpointSelectionStrategy;
+
+        StickySelector(ToLongFunction<ClientRequestContext> hasher, EndpointGroup endpointGroup,
+                       EndpointSelectionStrategy endpointSelectionStrategy) {
+            this.requestContextHasher = requireNonNull(hasher);
```

Comment:
Would you add a message parameter like `requireNonNull(hasher, "hasher")` to let a user know which parameter is `null`?